### PR TITLE
Llama 2 GQA

### PR DESCRIPTION
**Pull request by nrailgun**
Llama 2 GQA as described in paper 

https://arxiv.org/pdf/2305.13245.pdf
https://arxiv.org/abs/2307.09288

### CODE REVIEW

**Comment by nrailgun** [comment]
@microsoft-github-policy-service agree

**Comment by conglongli** [comment]
@nrailgun Thank you for the contribution. In order to make it easier for users to try this, could you also commit a Llama 2 pertaining script, something similar to https://github.com/microsoft/Megatron-DeepSpeed/blob/main/examples_deepspeed/pretrain_llama_distributed.sh? Thanks!

**Comment by nrailgun** [comment]
> @nrailgun Thank you for the contribution. In order to make it easier for users to try this, could you also commit a Llama 2 pretraining script, something similar to https://github.com/microsoft/Megatron-DeepSpeed/blob/main/examples_deepspeed/pretrain_llama_distributed.sh? Thanks!

No problem, I'll do it later.

File: examples_deepspeed/pretrain_llama2_distributed.sh [hunk 1]
```diff
@@ -0,0 +1,120 @@
+#!/bin/bash
+# This example script is contributed by external user https://github.com/LydiaXiaohongLi
```
Comment (conglongli) [inline examples_deepspeed/pretrain_llama2_distributed.sh:2]: Please change this credit to your own github account

File: megatron/arguments.py [hunk 2]
```diff
@@ -413,6 +413,17 @@ def validate_args(args, defaults={}):
...
+    # GQA
+    if args.num_key_value_heads is None:
+        args.num_key_value_heads = args.num_attention_heads
+    assert args.num_attention_heads % args.num_key_value_heads == 0, \
+        f"num_attention_heads must be divisible by num_key_value_heads (got `num_attention_heads`: {args.num_attention_heads} " \
+        f"and `num_key_value_heads`: {args.num_key_value_heads})."
+    if args.num_key_value_heads != args.num_attention_heads:
+        # if GQA
+        assert not args.mos, 'GQA currently does not support args.mos'
+        assert not args.kd, 'GQA currently does not args.kd'
```
Comment (conglongli) [inline megatron/arguments.py:425]: `does not args.kd` -> `does not support args.kd`

File: megatron/model/transformer.py [hunk 3]
```diff
@@ -474,15 +477,39 @@ def __init__(self, config, layer_number,
...
             config.num_attention_heads, world_size)
 
+        # Per GQA head and per partition values
+        kv_projection_size = config.kv_channels * config.num_key_value_heads
+        self.num_key_value_heads_per_partition = core.utils.divide(
+            config.num_key_value_heads, world_size)
+        self.num_key_value_groups = core.utils.divide(
+            config.num_attention_heads, config.num_key_value_heads)
+        assert self.hidden_size_per_attention_head == core.utils.divide(
+            kv_projection_size, config.num_key_value_heads)
```
Comment (conglongli) [inline megatron/model/transformer.py:487]: If these lines only apply to gqa, could you add a `if self.use_gqa:` condition?

**Review by conglongli — APPROVED** [review]
@nrailgun overall LGTM, I only left a few minor comments that should be easy to fix, and I will merge this PR after that, thank you.

**Comment by nrailgun** [comment]
> @nrailgun overall LGTM, I only left a few minor comments that should be easy to fix, and I will merge this PR after that, thank you.

DONE.

**Comment by zdaiot** [comment]
@nrailg Hello, how is the bookcorpus_train_1m_text_sentence dataset converted? 
In addition, how to convert the weight? 
Thanks a lot

**Comment by nrailg** [comment]
> @nrailg Hello, how is the bookcorpus_train_1m_text_sentence dataset converted? In addition, how to convert the weight? Thanks a lot

Nothing special. Just as described in README.md.

**Comment by zdaiot** [comment]
@nrailg Thank you for your reply, but I can't find the readme about examples_deepspeed/pretrain_llama2_distributed.sh. Could you send me a link?

**Pull request merged**